### PR TITLE
SCP-2566: Use new chain index effects everywhere in emulator, plutus-use-cases, plutus-pab and marlowe

### PR DESCRIPTION
**Pull request by koslambrou**
Used new chain index effects everywhere in emulator, plutus-use-cases and plutus-pab. Extended Contract API with new functions.

The old chain index functions are kept intact and functional for the time being. Therefore, there is A LOT of duplication, but it will be very soon removed and we will only keep the new chain index queries.

This PR is a WIP. There is still some refactoring left to do, but I wanted a quick review to see if I'm on the right path.

EDIT:

Here's what's left to do before merging:

- [x] Replace the use of `addressChangeRequest` in `Marlowe.Language.Client` since it uses the old chain index API. @nau 
- [x] Delete all of the functions with the `Old` which refer to the old chain index
- [x] Replace the chain index server in `plutus-pab` since we'll be using the one in `plutus-chain-index`.
- [x] Use `ChainIndexTx` instead of `OnChainTx` in `Plutus.Contract.Emulator.ContractInstance` and `Plutus.PAB.ContractInstance/STM` after old chain index is completely removed

Pre-submit checklist:
- Branch
    - [x] Commit sequence broadly makes sense
    - [x] Key commits have useful messages
    - [x] Relevant tickets are mentioned in commit messages
    - [x] Formatting, materialized Nix files, PNG optimization, etc. are updated
- PR
    - [x] Self-reviewed the diff
    - [x] Useful pull request description
    - [x] Reviewer requested

### CODE REVIEW

File: plutus-contract/src/Plutus/Contract.hs [hunk 1]
```diff
@@ -64,10 +79,14 @@ module Plutus.Contract(
...
     -- * Transactions
     , WalletAPIError
+    , Request.submitTxOld
     , Request.submitTx
     , Request.submitTxConfirmed
+    , Request.submitTxConstraintsOld
     , Request.submitTxConstraints
+    , Request.submitTxConstraintsSpendingOld
     , Request.submitTxConstraintsSpending
+    , Request.submitTxConstraintsWithOld
```
Comment (ak3n) [inline plutus-contract/src/Plutus/Contract.hs:89]: What about using of qualified imports and moving old functions to a different module instead of `Old` suffix?

**Review by j-mueller — COMMENTED** [review]
I thought we could have both chain index effects available everywhere, (by having requests for both of them in `PABReq`) instead of splitting everything into old vs new. So for the transition time we would have two chain index services available in the PAB and in the emulator. But still only one way of running traces in the emulator and contracts in the PAB.

File: plutus-contract/src/Plutus/Contract.hs (hunk 1 above)
Comment (koslambrou) [reply plutus-contract/src/Plutus/Contract.hs:93]: I'll be removing the functions with the `Old` suffix in this PR.

File: marlowe/src/Language/Marlowe/Client.hs [hunk 2]
```diff
@@ -152,29 +154,21 @@ instance Semigroup LastResult where
...
 
-    AddressChangeResponse{acrTxns} <- awaitPromise $ addressChangeRequest
-                AddressChangeRequest
-                { acreqSlotRangeFrom = 0
-                , acreqSlotRangeTo = slot
-                , acreqAddress = address
-                }
-    go acrTxns >>= checkpointLoop (follow client params)
+    go txs >>= checkpointLoop (follow client params)
+
```
[the remark this answers is not in the document]
Comment (j-mueller) [reply marlowe/src/Language/Marlowe/Client.hs:171]: We only need the list of marlowe inputs (`[Input]`), so you could change it to

```haskell
type MarloweFollowSchema = Endpoint "follow" ([Input], MarloweParams)
```

etc.

However, this is going to break the UI which calls these endpoints. So I think the best way forward for this PR is to just start the loop with `checkpointLoop (follow client params)` and delete the  `go txs >>=` bit.

[the remark this answers is not in the document]
Comment (j-mueller) [reply marlowe/src/Language/Marlowe/Client.hs:171]: @nau is going to add an alternative way of getting the history using the new chain index when this PR is merged

File: plutus-chain-index/src/Plutus/ChainIndex/Tx.hs [hunk 3]
```diff
@@ -19,38 +24,77 @@ module Plutus.ChainIndex.Tx(
...
     _citxData            :: Map DatumHash Datum,
+    -- ^ Datum objects recorded on this transaction.
     _citxRedeemers       :: Map RedeemerHash Redeemer,
+    -- ^ Redeemers of the minting scripts.
     _citxMintingPolicies :: Map MintingPolicyHash MintingPolicy,
+    -- ^ The scripts used to check minting conditions.
     _citxStakeValidators :: Map StakeValidatorHash StakeValidator,
-    _citxValidators      :: Map ValidatorHash Validator
+    _citxValidators      :: Map ValidatorHash Validator,
+    _citxIsValid         :: Bool
```
[the remark this answers is not in the document]
Comment (j-mueller) [reply plutus-chain-index/src/Plutus/ChainIndex/Tx.hs:63]: Ah yes, good idea.

File: marlowe/src/Language/Marlowe/Client.hs (hunk 2 above)
[the remark this answers is not in the document]
Comment (koslambrou) [reply marlowe/src/Language/Marlowe/Client.hs:171]: Alright, I'll revert back the Marlowe code.

File: plutus-use-cases/src/Plutus/Contracts/Auction.hs [hunk 4]
```diff
@@ -264,25 +263,43 @@ data BuyerEvent =
 waitForChange :: SlotConfig -> AuctionParams -> StateMachineClient AuctionState AuctionInput -> HighestBid -> Contract AuctionOutput BuyerSchema AuctionError BuyerEvent
 waitForChange slotCfg AuctionParams{apEndTime} client lastHighestBid = do
     t <- currentTime
+
+    let targetTime = TimeSlot.slotToBeginPOSIXTime slotCfg
```
Comment (j-mueller) [inline plutus-use-cases/src/Plutus/Contracts/Auction.hs:267]: can we use `apEndTime` as the target time? That way we don't need to loop as much.

(If this is a bigger change we could do it as a separate PR)

**Review by j-mueller — COMMENTED** [review]
Nice work! Just a few minor comments

File: marlowe/src/Language/Marlowe/Client.hs [hunk 5]
```diff
@@ -157,24 +157,16 @@ type MarloweContractState = LastResult
...
                 InProgress -> go rest
 
-    AddressChangeResponse{acrTxns} <- awaitPromise $ addressChangeRequest
-                AddressChangeRequest
-                { acreqSlotRangeFrom = 0
-                , acreqSlotRangeTo = slot
-                , acreqAddress = address
-                }
-    go acrTxns >>= checkpointLoop (follow client params)
+    go [] >>= checkpointLoop (follow client params)
```
[the remark this answers is not in the document]
Comment (koslambrou) [reply marlowe/src/Language/Marlowe/Client.hs:168]: I kept the `go` because if I deleted it, I would need to delete other unused functions like `updateHistoryFromTx` and `findInput`. However, I wanted to keep them for now and let Alex decide to remove it if necessary. Or I could comment them out. What do you think?

File: plutus-chain-index/src/Plutus/ChainIndex/Tx.hs [hunk 6]
```diff
@@ -19,44 +25,112 @@ module Plutus.ChainIndex.Tx(
...
-    _citxMintingPolicies :: Map MintingPolicyHash MintingPolicy,
-    _citxStakeValidators :: Map StakeValidatorHash StakeValidator,
-    _citxValidators      :: Map ValidatorHash Validator
+data ChainIndexTx =
+    ValidChainIndexTx {
+      _citxTxId            :: TxId,
+      -- ^ The id of this valid transaction.
+      _citxInputs          :: Set TxIn,
+      -- ^ The inputs to this valid transaction.
+      _citxOutputs         :: [TxOut],
```
[the remark this answers is not in the document]
Comment (koslambrou) [reply plutus-chain-index/src/Plutus/ChainIndex/Tx.hs:51]: I don't understand. I already made the change. There are 2 constructors now: `ValidChainIndexTx` and `InvalidChainIndexTx` where the difference is that the former has the `citxOutputs` field and latter doesn't have it.

[the remark this answers is not in the document]
Comment (koslambrou) [reply plutus-chain-index/src/Plutus/ChainIndex/Tx.hs:51]: Did you have something else in mind?

File: plutus-use-cases/src/Plutus/Contracts/Auction.hs (hunk 4 above)
Comment (koslambrou) [reply plutus-use-cases/src/Plutus/Contracts/Auction.hs:267]: Doing that results in an infinite loop. Recommend creating an issue for a separate PR.

File: plutus-chain-index/src/Plutus/ChainIndex/Tx.hs (hunk 6 above)
[the remark this answers is not in the document]
Comment (j-mueller) [reply plutus-chain-index/src/Plutus/ChainIndex/Tx.hs:51]: Yeah I thought it would be something like `_citxOutputs :: ChainIndexOutputs` with

```haskell
data ChainIndexOutputs =
  InvalidTx -- no outputs
  ValidTx [TxOut]
```

so that `ChainIndexTx` would only be a record

File: marlowe/src/Language/Marlowe/Client.hs (hunk 5 above)
[the remark this answers is not in the document]
Comment (j-mueller) [reply marlowe/src/Language/Marlowe/Client.hs:168]: ok, fair enough. @nau  can delete them if he wants.

File: plutus-chain-index/src/Plutus/ChainIndex/Tx.hs (hunk 6 above)
[the remark this answers is not in the document]
Comment (j-mueller) [reply plutus-chain-index/src/Plutus/ChainIndex/Tx.hs:51]: ok, sorry if that was unclear, I think I misunderstood what you were suggesting.

File: plutus-ledger/src/Ledger/Tx.hs [hunk 7]
```diff
@@ -14,25 +26,63 @@ module Ledger.Tx
...
+                          , _ciTxOutValue     :: Value
+                          }
+  deriving (Show, Eq, Serialise, Generic, ToJSON, FromJSON)
+
+makeLenses ''ChainIndexTxOut
+makePrisms ''ChainIndexTxOut
+
+-- | Converts a transaction output from the chain index to the plutus-ledger-api
+-- transaction output.
+toTxOut :: ChainIndexTxOut -> TxOut
```
[the remark this answers is not in the document]
Comment (koslambrou) [reply plutus-ledger/src/Ledger/Tx.hs:80]: Done. Also added the inverse property test between `fromTxOut/toTxOut`

File: plutus-chain-index/src/Plutus/ChainIndex/Tx.hs (hunk 6 above)
[the remark this answers is not in the document]
Comment (koslambrou) [reply plutus-chain-index/src/Plutus/ChainIndex/Tx.hs]: Done. Can you validate?

**Review by j-mueller — APPROVED** [review]
:shipit:
